### DISCUSSION TITLE
Change filter value before query

### DISCUSSION THREAD

**Discussion by dj-fiorex** [refinedev/refine#2946] [Q&A] [answered]
Hello to all, i have a simple question about how filter work.
Right now i use Supabase DataProvider, i have two table Operator and Operation. One operation has one operator.
This is the field operator in the operation list component:
```
{
    field: "operatorId",
    headerName: "Operator",
    type: "number",
    minWidth: 250,
    flex: 1,
    renderCell: function render({ row }) {
     if (operatorsIsLoading) {
      return "Loading...";
     }

     const operator = operatorsData?.data.find(
      (item) => item.id === row.operatorId
     );
     return operator?.name;
    },
},
```
Of course the DataGrid filter allow me to input an integer on the filter, how can i change this behaviour? I would like to show the operator name on the filter and then query the backend with the operator id.

Thanks to all

**Comment by omeraplak** [comment]
Hey @dj-fiorex ,
Sorry, I'm not sure I understood correctly, would it be helpful to add relational data to the table?

https://refine.dev/docs/advanced-tutorials/data-provider/supabase/#select---handling-one-to-many-relationship

**Comment by dj-fiorex** [comment, marked as answer]
Oh yes, i will try to explain better
This is my two table:
OPERATOR
| id<Int> | name | surname |
|----|------|------------|


OPERATIONS
| id | name | operatorId<Int> |
|----|------|------------|

This is the list component of the resource OPERATION
```typescript
export const OperationList: React.FC = () => {
    const { dataGridProps } = useDataGrid<IOperation>();
   
    const operatorIds = dataGridProps.rows.map((item) => item.operatorId);
    const { data: operatorsData, isLoading: operatorsIsLoading } =
     useMany<IOperator>({
      resource: "operators",
      ids: operatorIds,
      queryOptions: {
       enabled: operatorIds.length > 0,
      },
     });
   
   
    const columns = React.useMemo<GridColumns<IOperation>>(
     () => [
      { field: "id", headerName: "ID", flex: 1, minWidth: 50 },
      { field: "name", headerName: "Name", flex: 1, minWidth: 350 },
   
      {
       field: "operatorId",
       headerName: "Operator",
       type: "number",
       minWidth: 250,
       flex: 1,
       renderCell: function render({ row }) {
        if (operatorsIsLoading) {
         return "Loading...";
        }
   
        const operator = operatorsData?.data.find(
         (item) => item.id === row.operatorId
        );
        return operator?.name;
       },
      },
    
      {
       headerName: "Actions",
       field: "actions",
       minWidth: 250,
       renderCell: function render(params) {
        return (
         <Stack direction="row" spacing={1}>
          <EditButton hideText recordItemId={params.row.id} />
          <ShowButton hideText recordItemId={params.row.id} />
          <DeleteButton hideText recordItemId={params.row.id} />
         </Stack>
        );
       },
      },
     ],
     [
      operatorsIsLoading,
      operatorsData?.data,
     ]
    );
   
    return (
     <List
      cardHeaderProps={{
       title: "Operations",
       action: (
        <Stack direction="row" spacing={1}>
         <CreateButton />
         <ExportButton onClick={triggerExport} loading={exportLoading} />
        </Stack>
       ),
      }}
     >
      <DataGrid
       {...dataGridProps}
       columns={columns}
       autoHeight
       checkboxSelection
       disableSelectionOnClick
      />
     </List>
    );
   };
   

```

With this configuration the filter of the column operatorId will be of type NUMBER so i can select <, >, != ecc. This works ok because the query to the backend use the id.
What i want to do is to be able to filter the column operatorId by operator.name but i want to send operator.id to the backend

Thanks!

**Reply by salihozdemir** [reply]
Hey @dj-fiorex,
We understand your issue, In the Material UI documentation provides a solution in [here](https://mui.com/x/react-data-grid/column-definition/#special-properties). As soon as possible i'll create a new example for solution with refine.

Until then, you can like this:

```ts
{
     field: "operatorId",
     headerName: "Operator",
     type: "singleSelect",
     minWidth: 250,
     flex: 1,
     valueOptions: operatorsData?.data.map((operator) => ({ label: operator.name, value: operator.id })),
     renderCell: function render({ row }) {
      if (operatorsIsLoading) {
       return "Loading...";
      }
 
      const operator = operatorsData?.data.find(
       (item) => item.id === row.operatorId
      );
      return operator?.name;
     },
 },
```

If you need, you can customize the filter operators like [here](https://mui.com/x/react-data-grid/filtering/#customize-the-operators).

**Reply by dj-fiorex** [reply]
Hey @salihozdemir  first of all thanks for the response but i already tried this solution. In this way the value sent to backend is the operator.name and not the operator.id

**Reply by salihozdemir** [reply]
Could you try with `valueFormatter` like this:

```tsx
{
     field: "operatorId",
     headerName: "Operator",
     type: "singleSelect",
     minWidth: 250,
     flex: 1,
     valueOptions: operatorsData?.data.map((operator) => ({ label: operator.name, value: operator.id })),
     valueFormatter: (
          params: GridValueFormatterParams<{ label: string; value: number }>
        ) => {
          if (params.value == null) {
            return '';
          }

          return params.value;
       },
     renderCell: function render({ row }) {
      if (operatorsIsLoading) {
       return "Loading...";
      }
 
      const operator = operatorsData?.data.find(
       (item) => item.id === row.operatorId
      );
      return operator?.name;
     },
 },
```

**Reply by dj-fiorex** [reply]
Thank you @salihozdemir, now the filter work but there is some bug

- When i open the filter a call to the backend with an empty string is made
- when i select an operator from the filter i can not select another operator from the dropdown (see screen)

![Untitled](https://user-images.githubusercontent.com/15034741/199596101-4017186c-9025-4764-bb09-1d6a86b2aaf9.png)

Do guys have a clue on what's going on?

I will write some page of documentation and make a pr if it is welcome

Thanks

**Reply by dj-fiorex** [reply]
Tried something also, so if i override onFilterModelChange of the dataGridProps returned from useDataGrid<IOperation>() i can intercept the filter before it is sent to the server
```

onFilterModelChange: (
   model: GridFilterModel,
   details: GridCallbackDetails<"filter">
  ) => {
   console.log(model, details);
   if (!model.items[0].value) {
    model.items[0].value =
     model.items[0].columnField === "operatorId" ? operatorIds[1] : "1";
   }
   dataGridProps.onFilterModelChange(model, details);
  },

```

but i can not understand how can i block the request because there is no value.

Meanwhile i found another "bug":
When i select an operator from the filter (look at the screenshot of my last msg) the other operators does not show, this is because the useMany hook refresh and return only the selected operator, i think because there are no rows with other operators in the grid. Am i missing something?

**Reply by salihozdemir** [reply]
Hey @dj-fiorex, When you select a operator, your table data will update according to operatorId. Your `useMany` hook only fetch based on operators in table data.

So, you should change the your `useMany` hook to `useList` like below:

```tsx
const { data: operatorsData, isLoading: operatorsIsLoading } =
 useList<IOperator>({
  resource: "operators",
  config: {
      hasPagination: false,
    },
 });
```

**Reply by dj-fiorex** [reply]
Perfect, with the useList hook the filter is ok.
Remain another problem though, when i click on the cross icon to close the filter a call with an empty string is made to my backend.

Guys i have a lot of question about how to do things in refine, but i do not know where the best place for such questions is.

**Reply by omeraplak** [reply]
You can join our [Discord server](https://discord.gg/refine)

**Reply by salihozdemir** [reply]
> Perfect, with the useList hook the filter is ok. Remain another problem though, when i click on the cross icon to close the filter a call with an empty string is made to my backend.
> 
> Guys i have a lot of question about how to do things in refine, but i do not know where the best place for such questions is.

Hey @dj-fiorex,

I reproduced what you want to do and got a similar result. It must be a bug. Can you open an issue for this?
By the way, I noticed that it updates the filterModel and add filter with an empty string. So you see the empty select option. Until the issue is resolved, you can override `onFilterModelChange` like below:

```tsx
onFilterModelChange: (
   model: GridFilterModel,
   details: GridCallbackDetails<"filter">
  ) => {
   const filteredModel =  model.items.filter(item => item.value);
   dataGridProps.onFilterModelChange(filteredModel, details);
  },
```
To better understand what's going on, you can watch the `dataGridProps` with `console.log`.

**Reply by dj-fiorex** [reply]
Thanks a lot @salihozdemir 
issue opened https://github.com/refinedev/refine/issues/2992

By the way the solution that you suggested is not working because typescript give me an error "Argument of type 'GridFilterItem' is not assignable to parameter of type 'GridFilterModel'." i think because 
`const filteredModel =  model.items.filter(item => item.value)`
filteredModel is of type GridFilterItem and not GridFilterModel

EDIT:
maybe something like this works?
```
onFilterModelChange: (
   model: GridFilterModel,
   details: GridCallbackDetails<"filter">
  ) => {
   const filteredModel = {
    ...model,
    items: model.items.filter((item) => item.value),
   };
   dataGridProps.onFilterModelChange(filteredModel, details);
  },
```
